core/src/test/java/io/questdb/test/cairo/meta/BlockFileTest.java

```diff
@@ -0,0 +1,737 @@
+                    prevRegionLength = writer.getRegionLength(writer.getVersionVolatile());
+                }
+
+                // message C will not fit into the same region
+                try (BlockFileWriter writer = new BlockFileWriter(ff, commitMode)) {
+                    writer.of(path.$());
+                    int regionLength = BlockFileUtils.REGION_HEADER_SIZE;
+                    regionLength += commitMsgCVersion1(writer.append());
+                    regionLength += BlockFileUtils.BLOCK_HEADER_SIZE;
+                    writer.commit();
+                    assertRegionOffset(writer, 3, regionLength, prevRegionOffset + prevRegionLength);
+                }
+
+                readAllBlocks(path, 1);
+            }
+        });
+    }
+
+    @Test
+    public void testReadWriteChecksum() throws Exception {
+        assertMemoryLeak(() -> {
+            try (Path path = getDefinitionFilePath("test")) {
+                FilesFacade ff = configuration.getFilesFacade();
+                try (BlockFileWriter writer = new BlockFileWriter(ff, commitMode)) {
+                    writer.of(path.$());
+                    AppendableBlock memory1 = writer.append();
+                    commitMsgAVersion1(memory1);
+                    writer.commit();
+                }
+                readAllBlocks(path, 1);
+
+                // corrupt data
+                try (MemoryCMARW mem = Vm.getCMARWInstance()) {
+                    mem.smallFile(ff, path.$(), MemoryTag.MMAP_DEFAULT);
+                    mem.putInt(BlockFileUtils.HEADER_SIZE + BlockFileUtils.REGION_HEADER_SIZE + 8, -42);
+                }
+
+                try {
+                    readAllBlocks(path, 1);
+                    Assert.fail("checksum mismatch expected");
+                } catch (Exception e) {
+                    Assert.assertTrue(e.getMessage().contains("meta file checksum mismatch"));
+                }
+            }
+        });
+    }
+
+    @Test
+    public void testReadWriteConcurrently() throws Exception {
+        assertMemoryLeak(() -> {
+            FilesFacade ff = configuration.getFilesFacade();
+            Rnd rnd = TestUtils.generateRandom(LOG);
+            int readerThreads = 4;
+            CyclicBarrier start = new CyclicBarrier(readerThreads + 1);
+            AtomicInteger done = new AtomicInteger();
+            int iterations = 1000;
+            Thread writerThread = new Thread(() -> {
+                try (Path path = getDefinitionFilePath("test")) {
+                    start.await();
+                    for (int i = 0; i < iterations; i++) {
+                        int msg = rnd.nextInt(4);
+                        try (BlockFileWriter writer = new BlockFileWriter(ff, commitMode)) {
+                            writer.of(path.$());
+                            switch (msg) {
+                                case 0:
+                                    commitMsgAVersion1(writer.append());
```

Comment:
Please add a step that writes random variable length block, that changes it's size significantly. Say goes from 16 bytes to the 2 file page sizes to make sure the file does not grow in the size.

Reply:
The test has been updated.